src/Microsoft.SqlTools.ServiceLayer/EditData/Session.cs

```diff
@@ -0,0 +1,194 @@
+            EditCache = new ConcurrentDictionary<long, RowEditBase>();
+        }
+
+        #region Public Methods
+
+        /// <summary>
+        /// Validates that a query can be used for an edit session. The target result set is returned
+        /// </summary>
+        /// <param name="query">The query to validate</param>
+        /// <returns>The result set to use</returns>
+        public static ResultSet ValidateQueryForSession(Query query)
+        {
+            Validate.IsNotNull(nameof(query), query);
+
+            // Determine if the query is valid for editing
+            // Criterion 1) Query has finished executing
+            if (!query.HasExecuted)
+            {
+                throw new InvalidOperationException(SR.EditDataQueryNotCompleted);
+            }
+
+            // Criterion 2) Query only has a single result set
+            ResultSet[] queryResultSets = query.Batches.SelectMany(b => b.ResultSets).ToArray();
+            if (queryResultSets.Length != 1)
+            {
+                throw new InvalidOperationException(SR.EditDataQueryImproperResultSets);
+            }
+
+            return query.Batches[0].ResultSets[0];
+        }
+
+        /// <summary>
+        /// Creates a new row update and adds it to the update cache
+        /// </summary>
+        /// <exception cref="InvalidOperationException">If inserting into cache fails</exception>
+        /// <returns>The internal ID of the newly created row</returns>
+        public long CreateRow()
+        {
+            // Create a new row ID (atomically, since this could be accesses concurrently)
+            long newRowId = NextRowId++;
+
+            // Create a new row create update and add to the update cache
+            RowCreate newRow = new RowCreate(newRowId, associatedResultSet, objectMetadata);
+            if (!EditCache.TryAdd(newRowId, newRow))
+            {
+                // Revert the next row ID
+                NextRowId--;
+                throw new InvalidOperationException(SR.EditDataFailedAddRow);
+            }
+
+            return newRowId;
+        }
+
+        /// <summary>
+        /// Creates a delete row update and adds it to the update cache
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// If row requested to delete already has a pending change in the cache
+        /// </exception>
+        /// <param name="rowId">The internal ID of the row to delete</param>
+        public void DeleteRow(long rowId)
+        {
+            // Sanity check the row ID
+            if (rowId >= NextRowId || rowId < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(rowId), SR.EditDataRowOutOfRange);
+            }
+
+            // Create a new row delete update and add to cache
+            RowDelete deleteRow = new RowDelete(rowId, associatedResultSet, objectMetadata);
+            if (!EditCache.TryAdd(rowId, deleteRow))
```

Comment:
I was thinking that the UI should handle this and not offer the option to delete a row if there are changes pending for it. Sort of along the lines of "are you sure you want to exit without saving changes?" But I can easily swap this for a AddOrUpdate call.